File: content/develop/use-cases/prefetch-cache/go/cache.go
// Redis prefetch-cache helper.
//
// Each cached entity is stored as a Redis hash under cache:{prefix}:{id}
// with a long safety-net TTL that bounds memory if the sync pipeline
// ever stops, but is not the freshness mechanism. Freshness comes from
// the ApplyChange path, which the sync worker calls every time a
// primary mutation arrives.
//
// Reads run HGETALL against Redis only. A miss is not a fall-back
// trigger -- the application treats it as an error or a deliberate
// Invalidate for testing. In production a sustained miss rate means the
// prefetch or the sync pipeline is broken, not that the primary should
// be re-queried on the request path.
package prefetchcache

import (
	"context"
	"sort"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

// PrefetchCache is a prefetch-cache helper backed by Redis hashes with
// a safety-net TTL.
type PrefetchCache struct {
	client     *redis.Client
	prefix     string
	ttlSeconds int

	mu                sync.Mutex
	hits              int
	misses            int
	prefetched        int
	syncEventsApplied int
	syncLagMsTotal    float64
	syncLagSamples    int
}

// NewPrefetchCache returns a PrefetchCache. Pass an empty prefix to use
// the default "cache:category:" and 0 for ttlSeconds to use the default
// 3600.
func NewPrefetchCache(client *redis.Client, prefix string, ttlSeconds int) *PrefetchCache {
	if prefix == "" {
		prefix = "cache:category:"
	}
	if ttlSeconds == 0 {
		ttlSeconds = 3600
	}
	return &PrefetchCache{
		client:     client,
		prefix:     prefix,
		ttlSeconds: ttlSeconds,
	}
}

// Prefix returns the configured cache-key prefix.
func (c *PrefetchCache) Prefix() string { return c.prefix }

// TTLSeconds returns the configured safety-net TTL in seconds.
func (c *PrefetchCache) TTLSeconds() int { return c.ttlSeconds }

func (c *PrefetchCache) cacheKey(id string) string { return c.prefix + id }

func (c *PrefetchCache) stripPrefix(key string) string {
	if len(key) >= len(c.prefix) && key[:len(c.prefix)] == c.prefix {
		return key[len(c.prefix):]
	}
	return key
}

// BulkLoad pipelines DEL + HSET + EXPIRE for every record. Returns the
// number of records loaded.
//
// The pipeline is non-transactional: it is fast on startup (when
// nothing is reading the cache) and on the live /reprefetch path (when
// the demo pauses the sync worker around the call). Calling BulkLoad
// on a cache that is actively being read and written to can briefly
// expose a key that has been deleted but not yet rewritten; pause the
// writers first or rewrite this with TxPipeline if that matters.
func (c *PrefetchCache) BulkLoad(ctx context.Context, records []map[string]string) (int, error) {
	loaded := 0
	pipe := c.client.Pipeline()
	for _, record := range records {
		id := record["id"]
		if id == "" {
			continue
		}
		cacheKey := c.cacheKey(id)
		pipe.Del(ctx, cacheKey)
		fields := hashFields(record)
		pipe.HSet(ctx, cacheKey, fields...)
		pipe.Expire(ctx, cacheKey, time.Duration(c.ttlSeconds)*time.Second)
		loaded++
	}
	if loaded > 0 {
		if _, err := pipe.Exec(ctx); err != nil {
			return 0, err
		}
	}
	c.mu.Lock()
	c.prefetched += loaded
	c.mu.Unlock()
	return loaded, nil
}

// GetResult bundles the record, hit/miss flag, and Redis-side latency
// for a Get call.
type GetResult struct {
	Record         map[string]string
	Hit            bool
	RedisLatencyMs float64
}

// Get runs HGETALL against Redis and returns the cached hash with the
// hit flag and Redis-side latency in milliseconds.
//
// Prefetch-cache reads do not fall back to the primary. A miss is a
// signal that the cache is incomplete, not a trigger to re-query the
// source. The caller decides how to surface it.
func (c *PrefetchCache) Get(ctx context.Context, id string) (GetResult, error) {
	cacheKey := c.cacheKey(id)
	started := time.Now()
	cached, err := c.client.HGetAll(ctx, cacheKey).Result()
	latencyMs := float64(time.Since(started).Microseconds()) / 1000.0
	if err != nil {
		return GetResult{RedisLatencyMs: latencyMs}, err
	}
	if len(cached) > 0 {
		c.mu.Lock()
		c.hits++
		c.mu.Unlock()
		return GetResult{Record: cached, Hit: true, RedisLatencyMs: latencyMs}, nil
	}
	c.mu.Lock()
	c.misses++
	c.mu.Unlock()
	return GetResult{Record: nil, Hit: false, RedisLatencyMs: latencyMs}, nil
}

// ApplyChange applies a primary change event to Redis.
//
// For an upsert, the helper rewrites the cache hash and refreshes the
// safety-net TTL in one transactional pipeline so the cache never holds
// a stale mix of old and new fields. For a delete, it removes the cache
// key. An upsert with no fields is dropped silently: HSET with an empty
// mapping errors in most clients, and there is nothing to write.
func (c *PrefetchCache) ApplyChange(ctx context.Context, change Change) error {
	if change.ID == "" {
		return nil
	}
	cacheKey := c.cacheKey(change.ID)

	switch change.Op {
	case ChangeOpUpsert:
		if len(change.Fields) == 0 {
			// Malformed upsert with no fields. Skip rather than
			// crash the sync worker: HSET with an empty mapping
			// errors, and there's nothing to write anyway. A real
			// CDC consumer would route this to a dead-letter queue
			// and alert; the demo just drops it.
			return nil
		}
		pipe := c.client.TxPipeline()
		pipe.Del(ctx, cacheKey)
		pipe.HSet(ctx, cacheKey, hashFields(change.Fields)...)
		pipe.Expire(ctx, cacheKey, time.Duration(c.ttlSeconds)*time.Second)
		if _, err := pipe.Exec(ctx); err != nil {
			return err
		}
	case ChangeOpDelete:
		if err := c.client.Del(ctx, cacheKey).Err(); err != nil {
			return err
		}
	default:
		return nil
	}

	c.mu.Lock()
	c.syncEventsApplied++
	if change.TimestampMs > 0 {
		nowMs := float64(time.Now().UnixNano()) / 1e6
		lag := nowMs - change.TimestampMs
		if lag < 0 {
			lag = 0
		}
		c.syncLagMsTotal += lag
		c.syncLagSamples++
	}
	c.mu.Unlock()
	return nil
}

// Invalidate deletes one cache key. Returns true if a key was removed.
// Demo-only: simulates a broken sync pipeline.
func (c *PrefetchCache) Invalidate(ctx context.Context, id string) (bool, error) {
	n, err := c.client.Del(ctx, c.cacheKey(id)).Result()
	if err != nil {
		return false, err
	}
	return n == 1, nil
}

// Clear deletes every key under this cache's prefix using SCAN + DEL in
// batches. Returns the number of keys deleted.
func (c *PrefetchCache) Clear(ctx context.Context) (int, error) {
	var (
		cursor  uint64
		deleted int
	)
	for {
		keys, next, err := c.client.Scan(ctx, cursor, c.prefix+"*", 500).Result()
		if err != nil {
			return deleted, err
		}
		if len(keys) > 0 {
			n, err := c.client.Del(ctx, keys...).Result()
			if err != nil {
				return deleted, err
			}
			deleted += int(n)
		}
		cursor = next
		if cursor == 0 {
			break
		}
	}
	return deleted, nil
}

// IDs returns every entity ID currently in the cache, sorted, with the
// prefix stripped.
func (c *PrefetchCache) IDs(ctx context.Context) ([]string, error) {
	var (
		cursor uint64
		out    []string
	)
	for {
		keys, next, err := c.client.Scan(ctx, cursor, c.prefix+"*", 500).Result()
		if err != nil {
			return nil, err
		}
		for _, k := range keys {
			out = append(out, c.stripPrefix(k))
		}
		cursor = next
		if cursor == 0 {
			break
		}
	}
	sortStrings(out)
	return out, nil
}

// Count returns the number of keys under the cache prefix.
func (c *PrefetchCache) Count(ctx context.Context) (int, error) {
	var (
		cursor uint64
		count  int
	)
	for {
		keys, next, err := c.client.Scan(ctx, cursor, c.prefix+"*", 500).Result()
		if err != nil {
			return 0, err
		}
		count += len(keys)
		cursor = next
		if cursor == 0 {
			break
		}
	}
	return count, nil
}

// TTLRemaining returns the remaining TTL on the cached key in seconds
// (Redis TTL semantics: -2 = missing, -1 = no expiry).
//
// go-redis stores the sentinel -2/-1 values as raw nanosecond durations
// rather than multiplying by the seconds precision, so we have to
// translate them back to integer seconds explicitly.
func (c *PrefetchCache) TTLRemaining(ctx context.Context, id string) (int, error) {
	d, err := c.client.TTL(ctx, c.cacheKey(id)).Result()
	if err != nil {
		return 0, err
	}
	switch d {
	case -2 * time.Nanosecond:
		return -2, nil
	case -1 * time.Nanosecond:
		return -1, nil
	}
	return int(d.Seconds()), nil
}

// Stats returns the in-process counters and derived rates. JSON keys
// are snake_case to match the other client ports.
func (c *PrefetchCache) Stats() map[string]any {
	c.mu.Lock()
	defer c.mu.Unlock()
	total := c.hits + c.misses
	hitRate := 0.0
	if total > 0 {
		hitRate = roundTo(100.0*float64(c.hits)/float64(total), 1)
	}
	avgLag := 0.0
	if c.syncLagSamples > 0 {
		avgLag = roundTo(c.syncLagMsTotal/float64(c.syncLagSamples), 2)
	}
	return map[string]any{
		"hits":                c.hits,
		"misses":              c.misses,
		"hit_rate_pct":        hitRate,
		"prefetched":          c.prefetched,
		"sync_events_applied": c.syncEventsApplied,
		"sync_lag_ms_avg":     avgLag,
	}
}

// ResetStats zeroes every counter.
func (c *PrefetchCache) ResetStats() {
	c.mu.Lock()
	c.hits = 0
	c.misses = 0
	c.prefetched = 0
	c.syncEventsApplied = 0
	c.syncLagMsTotal = 0
	c.syncLagSamples = 0
	c.mu.Unlock()
}

// hashFields flattens a map into the [key1, val1, key2, val2, ...] slice
// go-redis expects for HSet.
func hashFields(record map[string]string) []any {
	fields := make([]any, 0, len(record)*2)
	for k, v := range record {
		fields = append(fields, k, v)
	}
	return fields
}

// roundTo rounds x to decimals digits after the decimal point.
func roundTo(x float64, decimals int) float64 {
	mul := 1.0
	for i := 0; i < decimals; i++ {
		mul *= 10
	}
	if x >= 0 {
		return float64(int64(x*mul+0.5)) / mul
	}
	return float64(int64(x*mul-0.5)) / mul
}

// sortStrings sorts a slice of IDs in place. Pulled out so callers can
// rely on a sorted result regardless of SCAN return order.
func sortStrings(s []string) {
	sort.Strings(s)
}

File: content/develop/use-cases/prefetch-cache/rust/cache.rs
//! Redis prefetch-cache helper.
//!
//! Each cached entity is stored as a Redis hash under
//! `cache:{prefix}:{id}` with a long safety-net TTL that bounds memory if
//! the sync pipeline ever stops, but is not the freshness mechanism.
//! Freshness comes from the `apply_change` path, which the sync worker
//! calls every time a primary mutation arrives.
//!
//! Reads run `HGETALL` against Redis only. A miss is not a fall-back
//! trigger — the application treats it as an error or a deliberate
//! `invalidate` for testing. In production a sustained miss rate means
//! the prefetch or the sync pipeline is broken, not that the primary
//! should be re-queried on the request path.

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::Instant;

use redis::{aio::ConnectionManager, AsyncCommands, AsyncIter, RedisResult};

use crate::primary::ChangeEvent;

#[derive(Debug, Clone)]
pub struct CacheConfig {
    pub prefix: String,
    pub ttl_seconds: i64,
}

impl Default for CacheConfig {
    fn default() -> Self {
        Self {
            prefix: "cache:category:".to_string(),
            ttl_seconds: 3600,
        }
    }
}

#[derive(Debug, Clone)]
pub struct GetResult {
    pub record: Option<HashMap<String, String>>,
    pub hit: bool,
    pub redis_latency_ms: f64,
}

#[derive(Default)]
struct CacheStats {
    hits: AtomicU64,
    misses: AtomicU64,
    prefetched: AtomicU64,
    sync_events_applied: AtomicU64,
    // sync lag is tracked as integer microseconds so we can use atomics
    // without an extra mutex. Converted to f64 ms at read time.
    sync_lag_us_total: AtomicU64,
    sync_lag_samples: AtomicU64,
}

#[derive(Clone)]
pub struct PrefetchCache {
    conn: ConnectionManager,
    cfg: CacheConfig,
    stats: Arc<CacheStats>,
}

impl PrefetchCache {
    pub fn new(conn: ConnectionManager, cfg: CacheConfig) -> Self {
        Self {
            conn,
            cfg,
            stats: Arc::new(CacheStats::default()),
        }
    }

    pub fn ttl_seconds(&self) -> i64 {
        self.cfg.ttl_seconds
    }

    #[allow(dead_code)]
    pub fn prefix(&self) -> &str {
        &self.cfg.prefix
    }

    fn cache_key(&self, id: &str) -> String {
        format!("{}{}", self.cfg.prefix, id)
    }

    fn strip_prefix<'a>(&self, key: &'a str) -> &'a str {
        key.strip_prefix(self.cfg.prefix.as_str()).unwrap_or(key)
    }

    /// Pipeline `DEL` + `HSET` + `EXPIRE` for every record. Returns the
    /// count loaded.
    ///
    /// The pipeline is non-transactional: it is fast on startup (when
    /// nothing is reading the cache) and on the live `/reprefetch` path
    /// (when the demo pauses the sync worker around the call). Calling
    /// `bulk_load` on a cache that is actively being read and written
    /// to can briefly expose a key that has been deleted but not yet
    /// rewritten; pause the writers first or rewrite this with an
    /// atomic pipeline if that matters.
    pub async fn bulk_load(
        &self,
        records: Vec<HashMap<String, String>>,
    ) -> RedisResult<u64> {
        let mut pipe = redis::pipe();
        let mut loaded: u64 = 0;
        for record in &records {
            let entity_id = match record.get("id") {
                Some(v) if !v.is_empty() => v.clone(),
                _ => continue,
            };
            let cache_key = self.cache_key(&entity_id);
            let pairs: Vec<(String, String)> =
                record.iter().map(|(k, v)| (k.clone(), v.clone())).collect();
            pipe.del(&cache_key).ignore();
            pipe.hset_multiple(&cache_key, &pairs).ignore();
            pipe.expire(&cache_key, self.cfg.ttl_seconds).ignore();
            loaded += 1;
        }
        if loaded > 0 {
            let mut conn = self.conn.clone();
            pipe.query_async::<_, ()>(&mut conn).await?;
        }
        self.stats.prefetched.fetch_add(loaded, Ordering::Relaxed);
        Ok(loaded)
    }

    /// Return `(record, hit, redis_latency_ms)` for an `HGETALL` against Redis.
    ///
    /// Prefetch-cache reads do not fall back to the primary. A miss is a
    /// signal that the cache is incomplete, not a trigger to re-query
    /// the source. The caller decides how to surface it.
    pub async fn get(&self, entity_id: &str) -> RedisResult<GetResult> {
        let cache_key = self.cache_key(entity_id);
        let mut conn = self.conn.clone();

        let started = Instant::now();
        let cached: HashMap<String, String> = conn.hgetall(&cache_key).await?;
        let redis_latency_ms = started.elapsed().as_secs_f64() * 1000.0;

        if !cached.is_empty() {
            self.stats.hits.fetch_add(1, Ordering::Relaxed);
            Ok(GetResult {
                record: Some(cached),
                hit: true,
                redis_latency_ms,
            })
        } else {
            self.stats.misses.fetch_add(1, Ordering::Relaxed);
            Ok(GetResult {
                record: None,
                hit: false,
                redis_latency_ms,
            })
        }
    }

    /// Apply a primary change event to Redis.
    ///
    /// For an `upsert`, rewrite the hash and refresh the safety-net TTL
    /// in a `MULTI`/`EXEC` pipeline. For a `delete`, remove the cache
    /// key. If `op == "upsert"` and `fields` is missing or empty,
    /// return early — calling `HSET` with no pairs raises in most
    /// clients, and there is nothing to write anyway.
    pub async fn apply_change(&self, change: &ChangeEvent) -> RedisResult<()> {
        if change.id.is_empty() {
            return Ok(());
        }
        let cache_key = self.cache_key(&change.id);
        let mut conn = self.conn.clone();

        match change.op.as_str() {
            "upsert" => {
                let fields = match &change.fields {
                    Some(f) if !f.is_empty() => f,
                    _ => {
                        // Malformed upsert with no fields. Skip rather
                        // than crash the sync worker. A real CDC
                        // consumer would route this to a dead-letter
                        // queue and alert; the demo just drops it.
                        return Ok(());
                    }
                };
                let pairs: Vec<(String, String)> =
                    fields.iter().map(|(k, v)| (k.clone(), v.clone())).collect();
                redis::pipe()
                    .atomic()
                    .del(&cache_key)
                    .ignore()
                    .hset_multiple(&cache_key, &pairs)
                    .ignore()
                    .expire(&cache_key, self.cfg.ttl_seconds)
                    .ignore()
                    .query_async::<_, ()>(&mut conn)
                    .await?;
            }
            "delete" => {
                let _: i64 = conn.del(&cache_key).await?;
            }
            _ => return Ok(()),
        }

        self.stats
            .sync_events_applied
            .fetch_add(1, Ordering::Relaxed);

        // Record sync lag: now_ms - change.timestamp_ms, clamped >= 0.
        let now_ms = now_unix_ms();
        let lag_ms = (now_ms - change.timestamp_ms).max(0.0);
        let lag_us = (lag_ms * 1000.0) as u64;
        self.stats
            .sync_lag_us_total
            .fetch_add(lag_us, Ordering::Relaxed);
        self.stats.sync_lag_samples.fetch_add(1, Ordering::Relaxed);
        Ok(())
    }

    /// Delete one cache key. Demo-only: simulates a broken sync pipeline.
    pub async fn invalidate(&self, entity_id: &str) -> RedisResult<bool> {
        let mut conn = self.conn.clone();
        let n: i64 = conn.del(self.cache_key(entity_id)).await?;
        Ok(n == 1)
    }

    /// Delete every key under this cache's prefix and return the count.
    pub async fn clear(&self) -> RedisResult<u64> {
        let mut conn = self.conn.clone();
        let pattern = format!("{}*", self.cfg.prefix);
        let mut keys: Vec<String> = Vec::new();
        {
            let mut iter: AsyncIter<String> = conn.scan_match(&pattern).await?;
            while let Some(key) = iter.next_item().await {
                keys.push(key);
            }
        }
        if keys.is_empty() {
            return Ok(0);
        }
        let mut deleted: u64 = 0;
        let mut conn = self.conn.clone();
        for chunk in keys.chunks(500) {
            let mut pipe = redis::pipe();
            for key in chunk {
                pipe.del(key);
            }
            let results: Vec<i64> = pipe.query_async(&mut conn).await?;
            deleted += results.into_iter().filter(|n| *n > 0).count() as u64;
        }
        Ok(deleted)
    }

    /// Return every entity id currently in the cache, sorted, prefix stripped.
    pub async fn ids(&self) -> RedisResult<Vec<String>> {
        let mut conn = self.conn.clone();
        let pattern = format!("{}*", self.cfg.prefix);
        let mut out: Vec<String> = Vec::new();
        let mut iter: AsyncIter<String> = conn.scan_match(&pattern).await?;
        while let Some(key) = iter.next_item().await {
            out.push(self.strip_prefix(&key).to_string());
        }
        out.sort();
        Ok(out)
    }

    pub async fn ttl_remaining(&self, entity_id: &str) -> RedisResult<i64> {
        let mut conn = self.conn.clone();
        Ok(conn.ttl(self.cache_key(entity_id)).await?)
    }

    pub fn stats(&self) -> serde_json::Value {
        let hits = self.stats.hits.load(Ordering::Relaxed);
        let misses = self.stats.misses.load(Ordering::Relaxed);
        let prefetched = self.stats.prefetched.load(Ordering::Relaxed);
        let sync_events_applied = self.stats.sync_events_applied.load(Ordering::Relaxed);
        let sync_lag_us_total = self.stats.sync_lag_us_total.load(Ordering::Relaxed);
        let sync_lag_samples = self.stats.sync_lag_samples.load(Ordering::Relaxed);
        let total = hits + misses;
        let hit_rate_pct = if total == 0 {
            0.0
        } else {
            ((1000 * hits / total) as f64) / 10.0
        };
        let sync_lag_ms_avg = if sync_lag_samples == 0 {
            0.0
        } else {
            let avg_us = sync_lag_us_total as f64 / sync_lag_samples as f64;
            (avg_us / 10.0).round() / 100.0
        };
        serde_json::json!({
            "hits": hits,
            "misses": misses,
            "hit_rate_pct": hit_rate_pct,
            "prefetched": prefetched,
            "sync_events_applied": sync_events_applied,
            "sync_lag_ms_avg": sync_lag_ms_avg,
        })
    }

    pub fn reset_stats(&self) {
        self.stats.hits.store(0, Ordering::Relaxed);
        self.stats.misses.store(0, Ordering::Relaxed);
        self.stats.prefetched.store(0, Ordering::Relaxed);
        self.stats.sync_events_applied.store(0, Ordering::Relaxed);
        self.stats.sync_lag_us_total.store(0, Ordering::Relaxed);
        self.stats.sync_lag_samples.store(0, Ordering::Relaxed);
    }
}

fn now_unix_ms() -> f64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64() * 1000.0)
        .unwrap_or(0.0)
}
